src/System.Drawing.Common/src/System/Drawing/Graphics.cs

```diff
@@ -20,6 +21,369 @@ public sealed partial class Graphics : MarshalByRefObject, IDisposable, IDeviceC
+                SafeNativeMethods.Gdip.CheckStatus(status);
+            }
+        }
+
+        public PixelOffsetMode PixelOffsetMode
+        {
+            get
+            {
+                PixelOffsetMode mode = 0;
+                int status = SafeNativeMethods.Gdip.GdipGetPixelOffsetMode(new HandleRef(this, NativeGraphics), out mode);
+                SafeNativeMethods.Gdip.CheckStatus(status);
+
+                return mode;
+            }
+            set
+            {
+                if (value < PixelOffsetMode.Invalid || value > PixelOffsetMode.Half)
+                {
+                    throw new InvalidEnumArgumentException(nameof(value), unchecked((int)value), typeof(PixelOffsetMode));
+                }
+
+                if (value == PixelOffsetMode.Invalid)
+                {
+                    throw new ArgumentException(SR.GdiplusInvalidParameter);
+                }
+
+                int status = SafeNativeMethods.Gdip.GdipSetPixelOffsetMode(new HandleRef(this, NativeGraphics), value);
+                SafeNativeMethods.Gdip.CheckStatus(status);
+            }
+        }
+
+        public Point RenderingOrigin
+        {
+            get
+            {
+                int x, y;
+                int status = SafeNativeMethods.Gdip.GdipGetRenderingOrigin(new HandleRef(this, NativeGraphics), out x, out y);
+                SafeNativeMethods.Gdip.CheckStatus(status);
+
+                return new Point(x, y);
+            }
+            set
+            {
+                int status = SafeNativeMethods.Gdip.GdipSetRenderingOrigin(new HandleRef(this, NativeGraphics), value.X, value.Y);
+                SafeNativeMethods.Gdip.CheckStatus(status);
+            }
+        }
+
+        public SmoothingMode SmoothingMode
+        {
+            get
+            {
+                SmoothingMode mode = 0;
+                int status = SafeNativeMethods.Gdip.GdipGetSmoothingMode(new HandleRef(this, NativeGraphics), out mode);
+                SafeNativeMethods.Gdip.CheckStatus(status);
+
+                return mode;
+            }
+            set
+            {
+                if (value < SmoothingMode.Invalid || value > SmoothingMode.AntiAlias)
+                {
+                    throw new InvalidEnumArgumentException(nameof(value), unchecked((int)value), typeof(SmoothingMode));
+                }
+
+                switch (value)
+                {
+                    case SmoothingMode.Default:
+                    case SmoothingMode.HighSpeed:
+                        value = SmoothingMode.None;
+                        break;
+                    case SmoothingMode.HighQuality:
+                        value = SmoothingMode.AntiAlias;
+                        break;
+                    case SmoothingMode.Invalid:
+                        throw new ArgumentException(SR.GdiplusInvalidParameter);
+                    default:
+                        break;
+                }
```

Comment:
Where did this switch block come from, and why?  It's not in the original Windows code; why is it needed here?

Reply:
This fixes the test `SmoothingMode_SetValid_GetReturnsExpected` with libgdiplus